### PR TITLE
BRIDGE-655 Add logic for setting a kill switch for unsupported client app versions

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java [hunk 1]
```diff
@@ -299,6 +303,30 @@ public boolean isHealthCodeExportEnabled() {
...
+    /** {@inheritDoc} */
+    @DynamoDBMarshalling(marshallerClass = StringIntegerMapMarshaller.class)
+    @Override
+    public Map<String,Integer> getMinSupportedAppVersions() {
+        return minSupportedAppVersions;
+    }
+
+    @Override
+    public void setMinSupportedAppVersions(Map<String,Integer> map) {
+        this.minSupportedAppVersions = (map == null) ? new HashMap<>() : map;
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:315]: It's set in constructor, so this.minSupportedAppVersions.addAll(map) would work here. Even if there were entries, they'd be updated.

File: app/org/sagebionetworks/bridge/dynamodb/StringIntegerMapMarshaller.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+import java.util.Map;
+
+import com.amazonaws.services.dynamodbv2.datamodeling.JsonMarshaller;
+
+/**
+ * Marshalls Map<String, Integer> values JSON when persisting using DynamoDB.
+ */
+public class StringIntegerMapMarshaller extends JsonMarshaller<Map<String,Integer>> {
+
+}
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/StringIntegerMapMarshaller.java]: What was your experience in needing this, because there's no implementation here. What did DynamoDB do without this?

File: app/org/sagebionetworks/bridge/models/ClientInfo.java [hunk 3]
```diff
@@ -134,6 +152,13 @@ public Integer getSdkVersion() {
         return sdkVersion;
     }
     
+    public boolean isSupportedVersion(Integer minSupportedVersion) {
+    	if (appVersion != null && minSupportedVersion != null && appVersion < minSupportedVersion) {
+    		return false;
+    	}
+    	return true;
+    }
+
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java:159]: return !(appVersion != null && minSupportedVersion != null && appVersion < minSupportedVersion);

File: app/org/sagebionetworks/bridge/models/ClientInfo.java [hunk 4]
```diff
@@ -78,6 +79,16 @@ public ClientInfo load(String userAgent) throws Exception {
...
+    
+    /**
+     * The osName that maps to an iOS client app.
+     */
+    public static final String IOS_OS_NAME = "iPhone OS";
+    
+    /**
+     * The osName that maps to an Android client app.
+     */
+    public static final String ANDROID_OS_NAME = "Android";
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java]: I don't see where these are used or how you're addressing the space vs "/" issue... ?

File: app/org/sagebionetworks/bridge/models/ClientInfo.java [hunk 5]
```diff
@@ -252,17 +284,52 @@ static ClientInfo parseUserAgentString(String ua) {
...
+        				withOsVersion(osParts[1].trim());
+        	}
+        	else {
+        		int idx = osInfo.lastIndexOf(" ");
+        		if (idx > 0) {
+            		builder = builder.
+            				withOsName(osInfo.substring(0,idx).trim()).
+            				withOsVersion(osInfo.substring(idx).trim());
+        		}
+        	}
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java]: Please replace tabs with four spaces.

File: app/org/sagebionetworks/bridge/models/studies/Study.java [hunk 6]
```diff
@@ -195,4 +196,19 @@
...
     public void setHealthCodeExportEnabled(boolean enabled);
+    
+    /**
+     * Minimum supported app version number. If set, user app clients pointing to an older version will 
+     * fail with an httpResponse status code of 410.
+     */
+    public Integer getMinSupportedAppVersion(String osName);
+    
+    /** @see #getMinSupportedVersion(); */
+    public void setMinSupportedAppVersion(String osName, Integer version);
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/studies/Study.java]: I would remove this setter, it's jut more common in Java to use getMinSupportedAppVersions().put(osName, version);

If you use it a lot, rename to addMinSupportedAppVersion(osName, version);

In reality this will be set with the map setter both by DynamoDB and the researcher UI, all as a group.

Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/models/studies/Study.java]: Really? seems verbose, but ok.

File: app/org/sagebionetworks/bridge/models/ClientInfo.java (hunk 3 above)
Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/models/ClientInfo.java:159]: yeah, I was trying to follow the style of the other method. Figured that the explicit "this is false and this is true" was preferred.

File: app/org/sagebionetworks/bridge/dynamodb/StringIntegerMapMarshaller.java (hunk 2 above)
Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/dynamodb/StringIntegerMapMarshaller.java]: Followed the previous example of StringSetMarshaller

File: app/org/sagebionetworks/bridge/play/controllers/BaseController.java [hunk 7]
```diff
@@ -160,6 +165,22 @@ private String getSessionToken() {
...
     }
+    
+    void verifySupportedVersionOrThrowException(Study study, UserSession session) throws UnsupportedVersionException {
+
+    	// Exit early if Admin. These accounts should not throw the exception b/c of version support
+    	// b/c might want to look at older data. The supported version is intended for client apps that
+    	// gather data rather than apps used to model it. syoung 11/19/2015
+    	if (session.isAdminRole()) {
+    		return;  
+    	}
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/play/controllers/BaseController.java]: I don't think we want to do this. These are orthogonal considerations and when we've made exceptions in the past for admin accounts, they've been used as participant accounts by developers and weird errors result.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java (hunk 1 above)
Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:315]: Is default constructor _always_ called? (not the case with iOS) Deserializer will use a different constructor from the base.

File: app/org/sagebionetworks/bridge/play/controllers/BaseController.java [hunk 8]
```diff
@@ -160,6 +165,22 @@ private String getSessionToken() {
...
+    	
+        ClientInfo clientInfo = getClientInfoFromUserAgentHeader();
+        String osName = clientInfo.getOsName();
+        if (osName != null && !clientInfo.isSupportedVersion(study.getMinSupportedAppVersion(osName))) {
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/play/controllers/BaseController.java]: null check unneeded. study.getMinSupportedAppVersion() will return null, null is checked for in isSupportedVersion. Which is good because if the osName is unknown, that method will return null (no map entry) as well.

File: app/org/sagebionetworks/bridge/services/StudyServiceImpl.java [hunk 9]
```diff
@@ -190,6 +190,7 @@ public Study updateStudy(Study study, boolean isAdminUpdate) {
         if (!isAdminUpdate) {
             study.setMaxNumOfParticipants(originalStudy.getMaxNumOfParticipants());
             study.setHealthCodeExportEnabled(originalStudy.isHealthCodeExportEnabled());
+            // TODO: syoung 11/16/2015 Should the minSupportedVersion require admin privileges? I'm thinking No.
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/StudyServiceImpl.java]: No, can remove comment.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoStudyTest.java [hunk 10]
```diff
@@ -37,34 +37,40 @@ public void equalsHashCode() {
...
+        assertEqualsAndNotNull(study.getConsentPDF(), JsonUtils.asText(node, "consentPDF"));
+        assertEqualsAndNotNull((Long)study.getVersion(), (Long)node.get("version").asLong());
         assertTrue(node.get("strictUploadValidationEnabled").asBoolean());
         assertTrue(node.get("healthCodeExportEnabled").asBoolean());
-        assertEquals("Study", node.get("type").asText());
+        assertEqualsAndNotNull("Study", node.get("type").asText());
+        
+        JsonNode supportedVersionsNode = JsonUtils.asJsonNode(node, "minSupportedAppVersions");
+        assertNotNull(supportedVersionsNode);
+        assertEqualsAndNotNull(study.getMinSupportedAppVersion("iPhone OS"), (Integer)supportedVersionsNode.get("iPhone OS").asInt());
```
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoStudyTest.java]: node.get("iPhone OS").asInteger() (or .asInt() I can't remember) -- no cast needed

File: app/org/sagebionetworks/bridge/play/controllers/AuthenticationController.java [hunk 11]
```diff
@@ -100,6 +102,7 @@ private Result signInWithRetry(final int retryCounter) throws Exception {
         final Study study = getStudyOrThrowException(json);
         try {
             session = authenticationService.signIn(study, signIn);
+            verifySupportedVersionOrThrowException(study, session);
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/play/controllers/AuthenticationController.java]: maybe  verifyMinSupportedAppVersionOrThrowException?

Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/play/controllers/AuthenticationController.java]: I think we should check this after line 95 as well. I don't know if this can happen, but if an application can be updated without losing its session, it would get back the session from Redis and we'd still want to block it at that time.

File: app/org/sagebionetworks/bridge/dynamodb/StringIntegerMapMarshaller.java (hunk 2 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/StringIntegerMapMarshaller.java]: You'll need to test this against DynamoDB. That marshaller is there because DDB has a weird thing where it will throw an error if an empty Set is set for that attribute. I don't know what it does with an empty Map... don't know if it can even serialize that by default (if not we'd need to serialize it as JSON). The DynamoStudyDaoTest runs against DDB so it can be verified to work there with an empty map, and with a map that has a min supported app version mapping.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java (hunk 1 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:315]: Yes, if there's no other constructor to call, Java calls the default constructor (always). If there's no default constructor... it adds one that does nothing in the byte code.

File: app/org/sagebionetworks/bridge/play/controllers/BaseController.java (hunk 8 above)
Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/play/controllers/BaseController.java]: you can ask for a null key in a HashMap? That throws an exception in iOS.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoStudyTest.java (hunk 10 above)
Comment (syoung-smallwisdom) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoStudyTest.java]: asInt() returns int, and not Integer. Needed to cast or it gave me an error.

File: app/org/sagebionetworks/bridge/play/controllers/BaseController.java (hunk 8 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/play/controllers/BaseController.java]: Not a null key, but if the key exists and the value doesn't, that's fine. Sorry, you might need the check that the key exists.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java (hunk 1 above)
Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:315]: addAll is different from set and doesn't meet the contract of the interface.

Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:315]: If you don't want to protect against an NPE (and that's not something we normally prevent), just set the map. It's just a little odd to create a new map in this situation. But because of how DDB works, we need to have this setter.

File: app/org/sagebionetworks/bridge/play/controllers/AuthenticationController.java (hunk 11 above)
Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/play/controllers/AuthenticationController.java]: Would like to keep the naming w/o the "min" since the exception it throws is "unsupported".

File: app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java (hunk 1 above)
Comment (syoung-smallwisdom) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:315]: Yeah, I get that. I was following the other examples of setters for things that shouldn't be set to null.

File: app/org/sagebionetworks/bridge/models/accounts/UserSession.java [hunk 12]
```diff
@@ -52,4 +53,7 @@ public User getUser() {
     public void setUser(User user) {
         this.user = user;
     }
+    public boolean isAdminRole() {
+    	return this.getUser().isInRole(Roles.ADMINISTRATIVE_ROLES);
+    }
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/accounts/UserSession.java]: I don't think this is called anywhere now and can be removed... ?

**Comment by syoung-smallwisdom** [comment]
closing until tests can be run. will reopen
